virtual_eeprom/virtual_eeprom.c

```diff
@@ -0,0 +1,227 @@
+#include <linux/uaccess.h>
+#include <linux/slab.h>
+#include <linux/fs.h>
+#include <linux/mutex.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
+#include <linux/list.h>
+
+// Structure to hold EEPROM data and metadata
+struct virtual_eeprom {
+    char *data;
+    struct proc_dir_entry *proc_entry;
+    struct mutex mutex;
+    int size;
+    const char *name;
+    // head for linked list management
+    struct list_head list;
+};
+
+// Function prototypes
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom);
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom);
+
+// Directory for the virtual EEPROMs
+static struct proc_dir_entry *eeprom_dir;
+// Initialize global list to track all EEPROMs
+static LIST_HEAD(eeprom_list);
+
+static ssize_t eeprom_read(struct file *file, char __user *buf, size_t count, loff_t *ppos)
+{
+    struct virtual_eeprom *eeprom = PDE_DATA(file_inode(file));
+    ssize_t ret;
+
+    if (*ppos >= eeprom->size)
+        return 0;
+
+    if (*ppos + count > eeprom->size)
+        count = eeprom->size - *ppos;
+
+    mutex_lock(&eeprom->mutex);
+    ret = copy_to_user(buf, eeprom->data + *ppos, count);
+    mutex_unlock(&eeprom->mutex);
+
+    if (ret == 0) {
+        *ppos += count;
+        return count;
+    } else {
+        return -EFAULT;
+    }
+}
+
+static ssize_t eeprom_write(struct file *file, const char __user *buf, size_t count, loff_t *ppos)
+{
+    struct virtual_eeprom *eeprom = PDE_DATA(file_inode(file));
+    int ret;
+
+    if (*ppos >= eeprom->size)
+        return -EFBIG;
+
+    if (*ppos + count > eeprom->size)
+        count = eeprom->size - *ppos;
+
+    mutex_lock(&eeprom->mutex);
+    if (copy_from_user(eeprom->data + *ppos, buf, count)) {
+        mutex_unlock(&eeprom->mutex);
+        return -EFAULT;
+    }
+
+    *ppos += count;
+    // Save the data to the file after each write
+    ret = save_eeprom_to_file(eeprom);
+    mutex_unlock(&eeprom->mutex);
+
+    return ret < 0 ? ret : count;
+}
+
+static const struct file_operations eeprom_fops = {
+    .owner = THIS_MODULE,
+    .read = eeprom_read,
+    .write = eeprom_write,
+};
+
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    mm_segment_t old_fs;
+    ssize_t ret;
+    char file_path[128];
+
+    snprintf(file_path, sizeof(file_path), "/mnt/settings/%s.bin", eeprom->name);
+
+    old_fs = get_fs();
+    set_fs(KERNEL_DS);
+    file = filp_open(file_path, O_WRONLY | O_CREAT, 0644);
+    if (IS_ERR(file)) {
+        set_fs(old_fs);
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_write(file, eeprom->data, eeprom->size, &file->f_pos);
+    filp_close(file, NULL);
+    set_fs(old_fs);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    mm_segment_t old_fs;
+    ssize_t ret;
+    char file_path[128];
+
+    snprintf(file_path, sizeof(file_path), "/mnt/settings/%s.bin", eeprom->name);
```

Comment:
Just need to be careful with load order, i.e. `/mnt/settings` needs to be mounted before initialising this module. You could defer loading the file until the first read?

Reply:
Thinking about this some more, wouldn't it be a better use of resources to always read from the file rather than keeping the entire file content in memory until the kernel module is unloaded?

Reply:
Also I think the file location should be pulled from the device tree to make the module super generic